lib/msf/core/db_manager/vuln.rb

```diff
@@ -31,17 +31,17 @@ def find_vuln_by_details(details_map, host, service=nil)
     vuln = nil
 
     if service
-      vuln = service.vulns.includes(:vuln_details).where(crit).first
+      other_vulns = service.vulns.includes(:vuln_details).where(crit)
+      vuln = other_vulns.empty? ? nil : other_vulns.first
```

Comment:
Just for context; What's the semantic difference here? 👀

Reply:
Rails 5 loses track of the db connection when `where` and `first` are combined here.  This eliminates the type confusion around that boundary.

```
msf5 > db_import ~/m3_report.xml
[*] Importing 'NeXpose XML Report' data
[*] Import: Parsing with 'Nokogiri v1.10.10'
[-] Error while running command db_import: No connection pool with 'primary' found.

Call stack:
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:1032:in `retrieve_connection'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_handling.rb:118:in `retrieve_connection'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_handling.rb:90:in `connection'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/arel-9.0.0/lib/arel/nodes/node.rb:49:in `to_sql'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql/schema_statements.rb:580:in `block in columns_for_distinct'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql/schema_statements.rb:578:in `map'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/postgresql/schema_statements.rb:578:in `columns_for_distinct'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/relation/finder_methods.rb:405:in `limited_ids_for'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/relation/finder_methods.rb:391:in `apply_join_dependency'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:550:in `block in exec_queries'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:584:in `skip_query_cache_if_necessary'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:547:in `exec_queries'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/association_relation.rb:34:in `exec_queries'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:422:in `load'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:200:in `records'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/relation.rb:195:in `to_ary'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/relation/finder_methods.rb:532:in `find_nth_with_limit'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/relation/finder_methods.rb:517:in `find_nth'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/relation/finder_methods.rb:125:in `first'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/msf/core/db_manager/vuln.rb:42:in `find_vuln_by_details'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/msf/core/db_manager/vuln.rb:182:in `block in report_vuln'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:416:in `with_connection'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/msf/core/db_manager/vuln.rb:96:in `report_vuln'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/metasploit/framework/data_service/proxy/vuln_data_proxy.rb:33:in `block in report_vuln'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/metasploit/framework/data_service/proxy/core.rb:166:in `data_service_operation'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/metasploit/framework/data_service/proxy/vuln_data_proxy.rb:31:in `report_vuln'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/rex/parser/nexpose_raw_nokogiri.rb:432:in `block in report_test'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:416:in `with_connection'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/rex/parser/nexpose_raw_nokogiri.rb:429:in `report_test'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/rex/parser/nexpose_raw_nokogiri.rb:100:in `end_element'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/nokogiri-1.10.10/lib/nokogiri/xml/sax/document.rb:127:in `end_element_namespace'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/nokogiri-1.10.10/lib/nokogiri/xml/sax/parser.rb:110:in `parse_with'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/nokogiri-1.10.10/lib/nokogiri/xml/sax/parser.rb:110:in `parse_memory'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/nokogiri-1.10.10/lib/nokogiri/xml/sax/parser.rb:83:in `parse'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/msf/core/db_manager/import/nexpose/raw.rb:12:in `import_nexpose_raw_noko_stream'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/msf/core/db_manager/import/nexpose/raw.rb:25:in `import_nexpose_rawxml'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/msf/core/db_manager/import.rb:100:in `import'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/msf/core/db_manager/import.rb:219:in `import_file'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/metasploit/framework/data_service/proxy/db_import_data_proxy.rb:17:in `block in import_file'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/metasploit/framework/data_service/proxy/core.rb:166:in `data_service_operation'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/metasploit/framework/data_service/proxy/db_import_data_proxy.rb:15:in `import_file'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1464:in `block (3 levels) in cmd_db_import'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1457:in `each'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1457:in `block (2 levels) in cmd_db_import'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1451:in `each'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1451:in `block in cmd_db_import'
/home/vagrant/.rvm/gems/ruby-2.6.6@metasploit-framework/gems/activerecord-5.2.4.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:416:in `with_connection'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/msf/ui/console/command_dispatcher/db.rb:1446:in `cmd_db_import'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:523:in `run_command'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:474:in `block in run_single'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:468:in `each'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:468:in `run_single'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/rex/ui/text/shell.rb:158:in `run'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
/home/vagrant/rapid7/metasploit/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'
./msfconsole:23:in `<main>'
```